crates/types/src/traits/block_contents.rs

```diff
@@ -50,7 +46,17 @@ pub trait BlockPayload:
     /// The type of the transitions we are applying
     type Transaction: Transaction;
 
-    // type Header: BlockHeader;
+    /// Data created during block building which feeds into the block header
+    type Metadata;
+
+    /// Build a payload and associated metadata with the transactions.
+    fn build(transactions: impl IntoIterator<Item = Self::Transaction>) -> (Self, Self::Metadata);
+
+    /// Encode the payload
+    fn encode(&self) -> Vec<u8>;
+
+    /// Decode the payload
+    fn decode(encoded_transactions: &[u8]) -> Self;
 
     /// returns hashes of all the transactions in this block
     /// TODO make this ordered with a vec
```

Comment:
Github won't let me comment on line 63 so I'll do it here. Does `transaction_commitments` method need to exist? If so then I think we want to change the return type so that it's no longer a `HashSet`.

Reply:
Yeah `transaction_commitments` is needed for the consensus and transactions tasks to update their `included_txns`, but it doesn't have to be a `HashSet`. I'll change it.

Reply:
Hmm. Why is hotshot using the `BlockPayload` trait to update their included txs? Seems like this task is not in the scope of `BlockPayload`.

Reply:
`transaction_commitments` is mainly used for metrics, e.g., `outstanding_transactions` and `outstanding_transactions_memory_size`. I could move this function to another trait since it's not in the scope of `BlockPayload`, but we still need to implement it as long as the metrics need it.

Is there any concern about having it in `BlockPayload` besides the extra implementation? Otherwise I'll probably just keep it there. I've changed it to a vector.